proof/refine/ARM/Ipc_R.thy

```diff
@@ -5223,26 +5274,268 @@ lemma replyPush_sch_act_wf:
+
+lemma blockedCancelIPC_st_tcb_at:
+  "\<lbrace>\<lambda>s. (t = t' \<longrightarrow> Q (P Inactive)) \<and> (t \<noteq> t' \<longrightarrow> Q (st_tcb_at' P t s))\<rbrace>
+   blockedCancelIPC st t' rptr
+   \<lbrace>\<lambda>_ s. Q (st_tcb_at' P t s)\<rbrace>"
+  unfolding blockedCancelIPC_def getBlockingObject_def
+  by (wpsimp wp: sts_st_tcb_at'_cases_strong replyUnlink_st_tcb_at' getEndpoint_wp hoare_vcg_imp_lift')
+
+lemma replyRemoveTCB_st_tcb_at'_cases:
+  "\<lbrace>\<lambda>s. (t = t' \<longrightarrow> Q (P Inactive)) \<and> (t \<noteq> t' \<longrightarrow> Q (st_tcb_at' P t s))\<rbrace>
+   replyRemoveTCB t'
+   \<lbrace>\<lambda>_ s. Q (st_tcb_at' P t s)\<rbrace>"
+  unfolding replyRemoveTCB_def cleanReply_def
+  by (wpsimp wp: replyUnlink_st_tcb_at' gts_wp' hoare_vcg_imp_lift')
+
+lemma cancelSignal_st_tcb_at'_cases:
+  "\<lbrace>\<lambda>s. (t = t' \<longrightarrow> Q (P Inactive)) \<and> (t \<noteq> t' \<longrightarrow> Q (st_tcb_at' P t s))\<rbrace>
+   cancelSignal t' n
+   \<lbrace>\<lambda>_ s. Q (st_tcb_at' P t s)\<rbrace>"
+  unfolding cancelSignal_def replyRemoveTCB_def cleanReply_def
+  by (wpsimp wp: sts_st_tcb_at'_cases_strong getNotification_wp hoare_vcg_imp_lift')
+
+lemma cancelIPC_st_tcb_at':
+  "\<lbrace>\<lambda>s. if t' = t \<and> st_tcb_at' (\<lambda>st. st \<notin> {Running, Restart, IdleThreadState}) t' s
+        then P (P' Inactive)
+        else P (st_tcb_at' P' t' s)\<rbrace>
+   cancelIPC t
+   \<lbrace>\<lambda>rv s. P (st_tcb_at' P' t' s)\<rbrace>"
+  apply (clarsimp simp: cancelIPC_def)
+  apply (rule hoare_seq_ext[OF _ stateAssert_sp])
+  apply (rule hoare_seq_ext[OF _ stateAssert_sp])
+  apply (rule hoare_seq_ext[OF _ gts_sp'])
+  apply (wpsimp wp: blockedCancelIPC_st_tcb_at replyRemoveTCB_st_tcb_at'_cases
+                    cancelSignal_st_tcb_at'_cases threadSet_pred_tcb_no_state)
+  apply (auto simp: pred_tcb_at'_def obj_at'_def)
+  done
+
+lemma cancelIPC_st_tcb_at'_different_thread:
+  "\<lbrace>\<lambda>s. P (st_tcb_at' st t' s) \<and> t \<noteq> t'\<rbrace> cancelIPC t \<lbrace>\<lambda>rv s. P (st_tcb_at' st t' s)\<rbrace>"
+  by (wpsimp wp: cancelIPC_st_tcb_at')
+
+lemma replyUnlink_replyTCB_None:
+  "\<lbrace>\<lambda>s. r \<noteq> rptr \<longrightarrow> obj_at' (\<lambda>reply. replyTCB reply = None) rptr s\<rbrace>
+   replyUnlink r t
+   \<lbrace>\<lambda>_. obj_at' (\<lambda>reply. replyTCB reply = None) rptr\<rbrace>"
+  apply (wpsimp wp: updateReply_wp_all gts_wp' simp: replyUnlink_def)
+  apply (auto simp: obj_at'_def projectKOs objBitsKO_def)
+  done
+
+lemma updateReply_replyPrev_replyTCB_None[wp]:
+  "updateReply r (replyPrev_update f) \<lbrace>obj_at' (\<lambda>reply. replyTCB reply = None) rptr\<rbrace>"
+  by (wp updateReply_wp_all) (auto simp: obj_at'_def projectKOs objBitsKO_def)
+
+lemma updateReply_replyNext_replyTCB_None[wp]:
+  "updateReply r (replyNext_update f) \<lbrace>obj_at' (\<lambda>reply. replyTCB reply = None) rptr\<rbrace>"
```

Comment:
There is `updateReply_obj_at’_inv` that should be able to handle these two cases.